### PR TITLE
Replace direct `cudaMemcpyAsync` calls with utility functions (limited to `cudf::io`)

### PR DESCRIPTION
**Pull request by vuule**
## Description

Issue https://github.com/rapidsai/cudf/issues/15620

Replaced the calls to `cudaMemcpyAsync` with the new `cuda_memcpy`/`cuda_memcpy_async` utility, which optionally avoids using the copy engine. Changes are limited to cuIO to make the PR easier to review (repetitive enough as-is!).

Also took the opportunity to use `cudf::detail::host_vector` and its factories to enable wider pinned memory use.

Skipped a few instances of `cudaMemcpyAsync`; few are under `io::comp`, which we don't want to invest in further (if possible). The other `cudaMemcpyAsync` instances are D2D copies, which `cuda_memcpy`/`cuda_memcpy_async` don't support. Perhaps they should, just to make the use ubiquitous.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/reader_impl_preprocess.cu [hunk 1]
```diff
@@ -218,7 +218,7 @@ void generate_depth_remappings(
  */
 [[nodiscard]] std::future<void> read_column_chunks_async(
   std::vector<std::unique_ptr<datasource>> const& sources,
-  std::vector<std::unique_ptr<datasource::buffer>>& page_data,
+  cudf::host_span<rmm::device_buffer> page_data,
```
Comment (vuule) [inline cpp/src/io/parquet/reader_impl_preprocess.cu:221]: simplified outdated complexity

File: cpp/src/io/orc/reader_impl_chunking.cu [hunk 2]
```diff
@@ -516,10 +516,10 @@ void reader_impl::load_next_stripe_data(read_mode mode)
         _stream.synchronize();
         stream_synchronized = true;
       }
-      device_read_tasks.push_back(
-        std::pair(source_ptr->device_read_async(
-                    read_info.offset, read_info.length, dst_base + read_info.dst_pos, _stream),
-                  read_info.length));
+      device_read_tasks.emplace_back(
```
Comment (vuule) [inline cpp/src/io/orc/reader_impl_chunking.cu:519]: Unrelated change; noticed clang-tidy complaining that we used to make an unnecessary move here :)

**Comment by ttnghia** [comment]
Do you have (run) any benchmark to make sure there is no regression?

**Comment by vuule** [comment]
> Do you have (run) any benchmark to make sure there is no regression?

I haven't because we currently don't do anything differently - we end up calling `cudaMemcpyAsync` on a pageable buffer.
I'll run all benchmarks once we actually move toward setting `allocate_host_as_pinned_threshold` and/or `kernel_pinned_copy_threshold`.

File: cpp/src/io/text/data_chunk_source_factories.cpp [hunk 3]
```diff
@@ -87,8 +87,10 @@ class datasource_chunk_reader : public data_chunk_reader {
       _source->host_read(_offset, read_size, reinterpret_cast<uint8_t*>(h_ticket.buffer.data()));
 
       // copy the host-pinned data on to device
-      CUDF_CUDA_TRY(cudaMemcpyAsync(
-        chunk.data(), h_ticket.buffer.data(), read_size, cudaMemcpyDefault, stream.value()));
+      cudf::detail::cuda_memcpy_async<char>(
```
Comment (pmattione-nvidia) [inline cpp/src/io/text/data_chunk_source_factories.cpp:90]: There are a number of places where the template argument (char) is given explcitly ... is the compiler really not able to deduce it from the inputs?

**Review by pmattione-nvidia — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/text/data_chunk_source_factories.cpp (hunk 3 above)
Comment (vuule) [reply cpp/src/io/text/data_chunk_source_factories.cpp:90]: hm, maybe it wasn't required here
In general, compiler can't do template type deduction + implicit conversion. So passing a container that get implicitly converted to a span requires the template type for `cuda_memcpy_async`.
